### ISSUE TITLE
mailwatch_milter_relay: very long execution time and high sql load

### ISSUE THREAD

**Issue by dneuhaeuser** [mailwatch/1.2.0#1265]
When running 'mailwatch_milter_relay.php' I observed that a run with '--refresh' (e.g. via cron) takes longer and longer, rising while mail.log is growing, from several minutes ultimatively to hours.
In this time there are very high loads on the mysql process.

When looking at the script I ask myself whether the code from line 143 onward really needs to be executed for each and every line in the logfile (each execution of function 'process_entries').
https://github.com/mailwatch/MailWatch/blob/ca1d81b4f2e5f788d70a84ae2daabc79cc330220/tools/Postfix_relay/mailwatch_milter_relay.php#L143

I believe it should probably be sufficient to run the codelines 143-163 only ONCE AFTER all logfile lines are processed.
All necessary informations for this are in the $idqueue array, right?

**Comment by ghost** [comment]
Hi @dneuhaeuser That would make for a good optimization, I believe.  This process has never been super efficient. I am open to ideas.

**Comment by dneuhaeuser** [comment]
I'm currently working on improving this scripts performance...

On an installation with about ~500 emails in the mail.log (filesize ~2,5 MB)
running mailwatch_milter_relay currently takes 36 minutes here with 100% cpu load on mysqld.

**Comment by ghost** [comment]
Just bear in mind under normal conditions --refresh should only be run once, preferrably after a logrotate even.  I look forward to your improvements.

**Comment by dneuhaeuser** [comment]
please look at the PR #1266 

it is a very simple change but effective:
the execution which took 36+ minutes before now completes in about 10 seconds.

probably there is still room for more optimization...

the cron file ('tools/Postfix_relay/mailwatch-milter-relay') executes the script hourly with --refresh

EDIT: there is actually a suggestion for the 'tail' mode in 'INSTALL.milter'
https://github.com/mailwatch/MailWatch/blob/ca1d81b4f2e5f788d70a84ae2daabc79cc330220/tools/Postfix_relay/INSTALL.milter#L33

I wonder why this method is not used in the suggested cron-file 
https://github.com/mailwatch/MailWatch/blob/ca1d81b4f2e5f788d70a84ae2daabc79cc330220/tools/Postfix_relay/mailwatch-milter-relay#L12

**Comment by ghost** [comment]
@dneuhaeuser you are right, we need to fix that discrepancy. I actually use a systemd unit for this.  It may make even more sense for us to include both a cron with the tail method and a systemd unit file that also uses the tail method.
